syntax = "proto3";
package coop.rchain.casper.protocol;

import "scalapb/scalapb.proto";
import "RhoTypes.proto";

option (scalapb.options) = {
  package_name: "coop.rchain.casper.protocol"
  flat_package: true
};

//-----------------------------------
//just a stub until we can do everything in rholang

message Bond {
  bytes validator = 1;
  int32 stake     = 2;
}

message RChainState {
  bytes tuplespace = 1; //hash of the tuplespace contents
  
  //Internals of what will be the "blessed" PoS contract 
  //(which will be part of the tuplespace in the real implementation).
  repeated Bond bonds        = 2;
  int64         blockNumber  = 3;
}

message NewTerm {
  Par term = 1;
}

message BondDiff {
  Bond added = 1;
}

message UnbondDiff {
  bytes validatorRemoved = 1;
}

message Diff {
  oneof diffClass {
    NewTerm    tdiff = 1;
    BondDiff   bdiff = 2;
    UnbondDiff udiff = 3;
  }
}

//------------------------------------

message Deploy {
  bytes user  = 1; //public key
  int32 nonce = 2; //for uniqueness
  Par   term  = 3; //rholang term to deploy
  bytes sig   = 4; //signature of (hash(resource) + nonce) using private key
}

service DeployService {
  rpc DoDeploy(Deploy) returns (DeployServiceResponse) {}
}
message DeployServiceResponse {
  bool success = 1;
}

message Justification {
  bytes validator       = 1;
  bytes latestBlockHash = 2;
}

message Header {
  repeated bytes parentsHashList = 1; //list of parent block hashes
  bytes postStateHash      = 2;
  bytes newCodeHash        = 3;
  bytes commReductionsHash = 4;
}

message Body {
  RChainState     postState      = 1; //will remove sending an explicit postState when history/rollback in tuplespace works
  repeated Deploy newCode        = 2; //new rholang code that was added to the state before processing
  repeated Diff   commReductions = 3; //the comm. rule reductions that occurred to obtain the post-state
}

//TODO: add fields for arbitrary data
message BlockMessage {
  bytes                  blockHash      = 1; //obtained by hashing the information in the header
  Header                 header         = 2;
  Body                   body           = 3;
  repeated Justification justifications = 4; // map of all validators to latest blocks based on current view
  bytes                  sender         = 5; //public key of the validator that created the block
  bytes                  sig            = 6; //signature generated by signing `hash(justification) + blockHash`.
}
